### PR TITLE
gtkd: 3.9.0 -> 3.10.0 and add possibility to use different D compilers

### PR DESCRIPTION
**Pull request by ShamrockLee**
###### Motivation for this change

*   Add the possibility to switch to other compilers:
    Specify the D compiler to use with argument `dcompiler` as some other PRs do. In addition, the attribute is pass-throughed to make it accessible for down-stream packages. (Continued work of #115265)

    *   ~~Possible future work: Fallback to `gdc` where `ldc` are not supported:
    `ldc` seems preferred by packagers here. However, the package currently supports only `x86_64-linux`, `i686-linux` and `x86_64-darwin`.
    Aarch64 build (with gdc) seems to encounter a bug in GDC that generate duplicate symbols and causes `multiple definition` errors. See
    https://github.com/gtkd-developers/GtkD/issues/282
    https://gcc.gnu.org/bugzilla/show_bug.cgi?id=92216
    After the fixed version get merged, we could make `gdc` as the fallback compiler option, so that `gtkd` would build on `aarch64-linux`.~~
    gtkd now compiles on `aarch64-linux` with `ldc`

*   Update 3.9.0 -> 3.10.0
    *   Remove the previously-specified patches
    *   Remove the `dev` output to resolve circular dependency between `out` and `dev` after update (Cc: @hedning )

*   Darwin build seems broken (probably due to the update).

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/gtkd/default.nix [hunk 1]
```diff
@@ -1,39 +1,24 @@
-{ lib, stdenv, fetchzip, fetchpatch, atk, cairo, ldc, gdk-pixbuf, gnome, gst_all_1, librsvg
+{ lib, stdenv, fetchzip, fetchpatch, atk, cairo, dcompiler, gdk-pixbuf, gnome, gst_all_1, librsvg
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/gtkd/default.nix:1]:
```suggestion
{ lib, stdenv, fetchzip, atk, cairo, dcompiler, gdk-pixbuf, gnome, gst_all_1, librsvg
```

File: pkgs/development/libraries/gtkd/default.nix [hunk 2]
```diff
@@ -1,39 +1,24 @@
...
 , glib, gtk3, gtksourceview4, libgda, libpeas, pango, pkg-config, which, vte }:
 
 let
   inherit (gst_all_1) gstreamer gst-plugins-base gst-plugins-bad;
 in stdenv.mkDerivation rec {
   pname = "gtkd";
-  version = "3.9.0";
-
-  outputs = [ "out" "dev" ];
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/gtkd/default.nix:10]: We most likely want to keep this.

Comment (ShamrockLee) [reply pkgs/development/libraries/gtkd/default.nix:10]:
```
error: cycle detected in build of '/nix/store/x6wh4v3hms7lnm0fm8xzxgh8y9ggi9m2-gtkd-3.10.0.drv' in the references of output 'dev' from output 'out'
```
Here's the full log:
[gtkd_761e87b336993b9fd7dd3c1bf64ac22162fad319.log](https://github.com/NixOS/nixpkgs/files/8657254/gtkd_761e87b336993b9fd7dd3c1bf64ac22162fad319.log)

File: pkgs/development/libraries/gtkd/default.nix [hunk 3]
```diff
@@ -1,39 +1,24 @@
...
-      sha256 = "0ngyqifw1kandc1vk01kms3z65pcisfd75q7z09rml96glhfzjd6";
-    })
-    # Fix breakage with dmd ldc 1.26 and newer
-    (fetchpatch {
-      url = "https://github.com/gtkd-developers/GtkD/commit/323ff96c648882eaca2faee170bd9e90c6e1e9c3.patch";
-      sha256 = "1rhyi0isl6fl5i6fgsinvgq6v72xq7c6sajrxcsnmrzpvw91il3d";
-    })
-  ];
-
   prePatch = ''
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/gtkd/default.nix:22]:
```suggestion
  postPatch = ''
```

File: pkgs/development/libraries/gtkd/default.nix (hunk 2 above)
Comment (SuperSandro2000) [reply pkgs/development/libraries/gtkd/default.nix:10]: hmmm, I don't know how to properly debug this.

File: pkgs/development/libraries/gtkd/default.nix (hunk 3 above)
Comment (ShamrockLee) [reply pkgs/development/libraries/gtkd/default.nix]: Rebased and squashed into the update commit.

**Pull request merged**